### ISSUE TITLE
[Feature Request] Transform that stacks data for agents with identical specs

### ISSUE THREAD

**Issue by kurtamohler** [pytorch/rl#2566]
## Motivation

Some multi-agent environments, like `VmasEnv`, stack all of the tensors for observations, rewards, etc. for different agents that have identical specs. For instance, in one of these stacked environments, if there are 2 agents that each have 8 observations, the observation spec might look like this:

```
Composite(
    group_0: Composite(
        observation: UnboundedContinuous(
            shape=torch.Size([2, 8]),
            ...),
        shape=torch.Size([2])),
    shape=torch.Size([]))
```

In contrast, other environments, like `UnityMLAgentsEnv`, have separate keys for each agent, even if the agents' specs are identical. For instance, with 2 agents that each have 8 observations, the observation spec might look like this:

```
Composite(
    group_0: Composite(
        agent_0: Composite(
            observation: UnboundedContinuous(
                shape=torch.Size([8]),
                ...),
            shape=torch.Size([])),
        agent_1: Composite(
            observation: UnboundedContinuous(
                shape=torch.Size([8]),
                ...),
            shape=torch.Size([])),
        shape=torch.Size([])),
    shape=torch.Size([]))
```

It is not easy to apply the same training script to two environments that use these two different formats. For instance, applying the [multi-agent PPO tutorial](https://pytorch.org/rl/main/tutorials/multiagent_ppo.html) to a Unity env is not straightforward.

## Solution

If we had an environment transform that could stack all the data from different keys, we could convert an environment that uses the unstacked format into an environment that uses the stacked format. Then it should be straightforward to use the same (or almost the same) training script on the two different environments.

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)

**Comment by thomasbbrunner** [comment]
Have you taken a look at the `group_map` argument? When set to `MarlGroupMapType.ALL_IN_ONE_GROUP` the environment should return all agents in a single group (when possible, otherwise in more than one group.

If you are using this, then imo this is more of an issue with the implementation of the grouping of agents in `UnityMLAgentsEnv`.

**Comment by kurtamohler** [comment]
@thomasbbrunner, the `group_map=ALL_IN_ONE_GROUP` argument does combine all the agents into one group key, but the specific issue is that the data for each of the agents do not get stacked all into the same set of tensors--each agent still has a separate key containing separate tensors, like the example I showed in the issue description.

One of the reasons why `UnityMLAgentsEnv` does not stack the tensors for all the agents is that each of the agents in a Unity env are free to have different specs--for instance, the strikers and goalies have different specs in the "StrikersVsGoalies" example environment. In cases like that, it's not possible to stack all of the agents' data together. So when I was writing the wrapper, it seemed natural to keep the agents under separate agent keys, rather than try to automatically figure out which agents' specs match each other and can be stacked together. Furthermore, it seemed best to organize the data in a way that is consistent with how the `ml_agents` library itself does it.

But I'm happy to change the behavior if there is good reason to do so.

**Comment by thomasbbrunner** [comment]
Ah, I see! Thanks for the explanation. It seems that the `UnityMLAgentsEnv` has a different behavior than the other multi-agent environments (my experience is mostly based on PettingZoo here).

`PettingZooEnv` has three possible grouping types:
1. `ONE_GROUP_PER_AGENT` (self explanatory)
2. `None` (default): will group as many agents as possible, while keeping the groups homogeneous (i.e., all agents in the group have the same obs and action spec).
3. `ALL_IN_ONE_GROUP`: (self explanatory) + if the agents cannot be grouped together (due to mismatched specs), it will return some weird spec and any environment operation will lead to an error. 

Personally, I don't see the benefit of grouping heterogeneous agents. Maybe that makes sense in your use-case, but I'd argue that there is no difference between having one group for each agent and a group containing sub-groups of heterogenous agents.

I quite like the default behavior of `PettingZooEnv` (grouping where possible) and I don't understand why there isn't a `GROUP_HOMOGENEOUS` (or similar) in `MarlGroupMapType`.

Would be interested in hearing more about your use-case!

**Comment by thomasbbrunner** [comment]
It would be nice if the behavior of all TorchRL environments would be consistent with each other. It sometimes feels like multi-agent is not part of the "core" TorchRL capabilities and I'm hoping we can change that!

**Comment by kurtamohler** [comment]
I agree that we should make environments consistent with each other when possible. At the time, I didn't think that putting the Unity env agents under separate keys would be a significant inconsistency with other TorchRL envs  (at least compared to VMAS, PettingZoo, and OpenSpiel)--it just seemed like the right choice because it's more consistent with the underlying `ml_agents` library. But I see that it probably would be better to stack the data of agents that share the same group, by default.

@vmoens, do you agree with this?

> I don't understand why there isn't a `GROUP_HOMOGENEOUS`

That sounds like a good idea to me. Feel free to submit an issue

> Personally, I don't see the benefit of grouping heterogeneous agents.

I don't either, but then again, I'm fairly new here. Maybe I took `ALL_IN_ONE_GROUP` too literally

**Comment by vmoens** [comment]
> It would be nice if the behavior of all TorchRL environments would be consistent with each other. It sometimes feels like multi-agent is not part of the "core" TorchRL capabilities and I'm hoping we can change that!

I'm sorry this is the impression you have, @matteobettini gave a great deal of effort in unifying MARL APIs, but if there are inconsistencies we should address them!

> One of the reasons why UnityMLAgentsEnv does not stack the tensors for all the agents is that each of the agents in a Unity env are free to have different specs--for instance, the strikers and goalies have different specs in the "StrikersVsGoalies" example environment. In cases like that, it's not possible to stack all of the agents' data together. So when I was writing the wrapper, it seemed natural to keep the agents under separate agent keys, rather than try to automatically figure out which agents' specs match each other and can be stacked together. Furthermore, it seemed best to organize the data in a way that is consistent with how the ml_agents library itself does it.

I 100% agree on this.
My take when writing wrappers is usually that I want the wrapper to have the minimal interference with the underlying API – the only thing we want is to present the data in a tensordict-friendly manner: ie, convert to tensors, make sure that the data is on the desired device, minimize the cost of operations etc. Once that is done, we leave it to the user to deal with the data. 
The reason we want to do this that any change we make in the way the data is presented to the user will create edge cases that we may not know of. 

Now RE the "consistentcy" problem highlighted by @thomasbbrunner, as I said earlier we should make sure that things can be made uniform at a low cost.
My vision here would be for TorchRL to deal itself with the various grouping strategies with its own API, possibly inspired by pettingzoo if that is a convention in the community.
The API would consist in appending a transform that re-arranges the tensordict given some flag, for instance you could create your Unity env in its wrapper and pass a transform
```python
env = UnityMLAgentsEnv(...)
transform = GroupMARLAgents(MARLGroups.ONE_GROUP_PER_AGENT)
env = env.append_transform(transform)
```
and then, internally, we make sure that the transform does its job for every MARL library we support.

Would that make sense?

nb: Using metaclasses, we could even pass directly a group argument in each wrapper that automatically append the transform if required.

**Comment by matteobettini** [comment]
Maybe i am missing what the `group_map` input to the MLAgents constructor does. Cause that is what I would use to change the data grouping and address this issue

**Comment by kurtamohler** [comment]
I think I understand what you're saying, @matteobettini. I think two main points to summarize this discussion are:

1. By default, UnityMLAgentsEnv should not stack agent data, and the structure of the specs should reflect the group IDs assigned by MLAgents. (At least Vincent and I seem to agree on this)
2. All agents that are in the same MARL group should always be stacked together.

An important distinction is that the MLAgents group IDs are not the same thing as MARL group keys.

At the moment, UnityMLAgentsEnv honors point 1, but not point 2. So how do we honor both points?

Consider an example unity env where there are two unity group IDs, which each have 3 agents.

Right now, the default group map that UnityMLAgentsEnv gives would be this:

```
{
    'group_0': ['agent_0', 'agent_1', 'agent_2'],
    'group_1': ['agent_3', 'agent_4', 'agent_5'],
}
```

In order to honor both points 1 and 2, we would need to change UnityMLAgentsEnv's default group map to:

```
{
    ('group_0', 'agent_0'): ['agent_0'],
    ('group_0', 'agent_1'): ['agent_1'],
    ('group_0', 'agent_2'): ['agent_2'],
    ('group_1', 'agent_3'): ['agent_3'],
    ('group_1', 'agent_4'): ['agent_4'],
    ('group_1', 'agent_5'): ['agent_5'],
}
```

And also, to allow the user to specify a group map where some agents are in the same MARL group, UnityMLAgentsEnv needs to know how to stack/unstack the data coming out of and going into MLAgents.

Does this make sense to everyone?

If so, then the next question is whether the `Stack` transform would still be useful. I suppose it could--maybe a user wants to have fine-grain control over which data fields get stacked and which are kept unstacked

**Comment by matteobettini** [comment]
> Isn't only axiom 2 violated? The specs and tensordicts currently do match for the MLAgents wrapper. That is, unless you've seen a bugged case where that's not true. As far as I understand, the only thing that MLAgents violates is the part of axiom 2 that says the specs and tensordicts in a group must be stacked.

Nice! I think I had understood wrongly

> Is my solution https://github.com/pytorch/rl/issues/2566#issuecomment-2499296045 insufficient in some way?

I think it makes sense but I am not sure I understood it fully. I think it is because I have not clear what the `group_map` arg currently does.

Two points about that solution that I am still wondering about:
- isn't `group_map == mlagents_group_map` the best default value? Why are we having one group per agent as default?
- The names of the groups cannot be tuples (this is just a nit)

It is totally fine if you want to remove completely the group_map arg from MLagents, but I personally think it is a nice feature to have and it aligns with the many envs already present.

There is also a point about the transform adding an overhead which we can talk about today in our meeting

**Comment by matteobettini** [comment]
> A further point is that if we used a transform for all MARL groupings, rather than requiring the env wrappers to implement the MARL grouping logic, then the implementation of env wrappers would become a bit simpler and less prone to potentially inefficient or bugged implementations of the MARL grouping logic. It seems like it would be nice to have the MARL grouping logic implemented in one common place.

I agree with this but I think it could be done in a utility function (which many MARL envs can use) rather than a transform.

Since all MARL envs already provide this functionality and it is now part of the MARL API, I don't see why removing it. Instead we could consider making a utli function to help in providing it.

**Comment by matteobettini** [comment]
Let me take a concrete example.

This is how VMAS stacks the data coming from the step output into the groups dictated by the `group_map`
https://github.com/pytorch/rl/blob/d537dcb6347e2370fcdaed553bf3474d653cb5a6/torchrl/envs/libs/vmas.py#L577-L608

The original vmas interface just outputs lists  (e.g., `obs[i]` is the obs of agent i).

Can you see how doing the flexible stacking at this stage is better than creating a tensordict for each individual agent, outputting it, and doing the stacking later with another for loop over the agents?

The amout of code you would have if you do the "one td per agent" solution would be the exact same, but it would just be more inefficient
